corehq/apps/linked_domain/updates.py

```diff
@@ -173,16 +174,31 @@ def update_user_roles(domain_link):
         reduce=False,
     )
     local_roles_by_name = {role.name: role for role in local_roles}
+
+    # Update downstream roles based on upstream roles
     for role_def in master_results:
         role = local_roles_by_name.get(role_def['name'])
         if role:
             role_json = role.to_json()
         else:
             role_json = {'domain': domain_link.linked_domain}
 
+        role_def = copy(role_def)
+        role_def.pop('_id')
         role_json.update(role_def)
+        local_roles_by_name[role_json['name']] = role_json
         UserRole.wrap(role_json).save()
 
+    # Update assignable_by ids - must be done after main update to guarantee all local roles have ids
+    master_roles_by_id = {role['_id']: role for role in master_results}
+    for role in local_roles_by_name.values():
+        if role['assignable_by']:
+            role['assignable_by'] = [
+                local_roles_by_name[master_roles_by_id[role_id]['name']]['_id']
```

Comment:
Confirming that we dont need to fetch local_roles_by_name again so that they have `_id`?

Reply:
Yeah, `_id` gets assigned to the dicts when they're saved. The test should cover this.

Reply:
oh yes, tests did cover this but i was not sure how that is even working. thank you.

Reply:
Actually, i just checked that we allow creating two roles with the same name. Which would get tricky here. It also made the UI funky (notice the missing "delete" button)
<img width="1024" alt="Screenshot 2020-07-29 at 1 53 53 PM" src="https://user-images.githubusercontent.com/3864163/88775943-f8f60000-d1a2-11ea-9f2b-763b0be8dc53.png">

Reply:
Yeah, linked domains relies on names, which is a risk, but I don't think this PR introduces any new problems or makes existing problems worse.

Reply:
> linked domains relies on names, which is a risk

hmm, that is definitely not something I think we should ignore. Is there something we plan to do about it?

Reply:
Added https://trello.com/c/FJWf2qPH/122-make-linked-domains-less-name-dependent

I think it'd be relatively small to fix, maybe 2 days of work including tests and backfilling existing roles.

Reply:
thank you. I was thinking the impact of this could be a user getting access to a role and hence permissions the domain admin didn't intend to give them leading to security issues.
I think we could just not allow having roles with same name. That looks like worth considering as the solution anyway.